Fix missing InputEventMIDI from MIDI with multiple messages eg chord

markofjohnson opened the pull request:
Fix missing InputEventMidi from midi with multiple messages eg chord

MIDIDriver::receive_input_packet can be called to process an incoming midi packet that contains multiple messages. Eg playing a chord (several notes at the same time) may arrive using midi 'running status' and contain several NoteOn in one packet. This commit loops through all the messages in the midi packet creating an InputEventMidi for each one.

Fixes #77035

Simple fix but quite a serious MIDI bug that can be reproduced with eg Roland keyboards connected to macOS by bluetooth. Could effect Windows & Linux too, reproducibility will depend on hardware and OS midi configuration. Suggest consideration for Godot 4.2 and 3.x as their midi_driver.cpp is almost the same and has the same bug. Thanks.

<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

lawnjelly commented:
You will need to squash commits:
https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html#the-interactive-rebase

There will probably be a few nitpicks but the principle looks sound, well done for spotting this.

AThousandShips commented on core/os/midi_driver.cpp:
```diff
@@ -43,76 +43,82 @@ void MIDIDriver::set_singleton() {
-			last_received_message = data[0];
-		}
-	}
+	// *data may contain multiple messages, eg multiple note on using midi 'running status'
```
```suggestion
	// data may contain multiple messages, eg multiple note on using midi 'running status'.
```
Per the [comment style](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#comment-style-guide).

AThousandShips commented on core/os/midi_driver.cpp:
```diff
@@ -43,76 +43,82 @@ void MIDIDriver::set_singleton() {
-	}
+	// *data may contain multiple messages, eg multiple note on using midi 'running status'
+	//printf("MIDIDriver::receive_input_packet length %u\n",length);
+	uint32_t index = 0; // index will be incremented as each status or data byte is read from *data
```
```suggestion
	uint32_t index = 0; // index will be incremented as each status or data byte is read from data.
```

AThousandShips commented on core/os/midi_driver.cpp:
```diff
@@ -43,76 +43,82 @@ void MIDIDriver::set_singleton() {
-				event->set_controller_value(data[param_position + 1]);
+		if (length - index >= 1) {
+			if (data[index] >= 0xF0) {
+				// channel does not apply to system common messages
```
```suggestion
				// Channel does not apply to system common messages.
```

AThousandShips commented on core/os/midi_driver.cpp:
```diff
@@ -43,76 +43,82 @@ void MIDIDriver::set_singleton() {
+				last_received_message = data[index];
+				index++;
+			} else if ((data[index] & 0x80) == 0x00) {
+				// running status, index is not changed because no new midi status byte, just additional data bytes
```
```suggestion
				// Running status, index is not changed because no new MIDI status byte, just additional data bytes.
```

AThousandShips commented on core/os/midi_driver.cpp:
```diff
@@ -43,76 +43,82 @@ void MIDIDriver::set_singleton() {
+			case MIDIMessage::PITCH_BEND:
+				if (length >= 2 + index) {
+					event->set_pitch((data[index] << 7) | data[index]);
+					index++; // clearer than trying to increment on line above given the or operator
```
```suggestion
					index++;
```
Unnecessary

AThousandShips commented on core/os/midi_driver.cpp:
```diff
@@ -43,76 +43,82 @@ void MIDIDriver::set_singleton() {
+		}
+
+		Input *id = Input::get_singleton();
+		id->parse_input_event(event); // new InputEventMidi for each message read from *data
```
```suggestion
		Input::get_singleton()->parse_input_event(event);
```
No need to fetch as a variable for one line, comment unnecessary

AThousandShips commented on core/os/midi_driver.cpp:
```diff
@@ -43,76 +43,82 @@ void MIDIDriver::set_singleton() {
-		}
-	}
+	// *data may contain multiple messages, eg multiple note on using midi 'running status'
+	//printf("MIDIDriver::receive_input_packet length %u\n",length);
```
```suggestion
```

AThousandShips commented:
Also please make your one commit have a clear message, like that of the PR title

AThousandShips commented on core/os/midi_driver.cpp:
```diff
@@ -43,76 +43,82 @@ void MIDIDriver::set_singleton() {
 }
 
 void MIDIDriver::receive_input_packet(int device_index, uint64_t timestamp, uint8_t *data, uint32_t length) {
```
Since we're rewriting this let's fix the arguments
```suggestion
void MIDIDriver::receive_input_packet(int p_device_index, uint64_t p_timestamp, uint8_t *p_data, uint32_t p_length) {
```

markofjohnson commented:
Hey @AThousandShips I'm struggling with git. Is it in the correct state now? 

Also, there is another change I will propose to this method to protect against a hang if a malformed midi packet comes in (which given it could come from any MIDI device is possible). How should I proceed? I am thinking I will make another commit and ask for feedback?

AThousandShips commented:
You can add it as a separate commit to make it easy to remove if it's not desired :) the commit history looks good now

lawnjelly commented on core/os/midi_driver.cpp:
```diff
@@ -42,77 +42,82 @@ void MIDIDriver::set_singleton() {
-			if (length >= 2 + param_position) {
-				event->set_controller_number(data[param_position]);
-				event->set_controller_value(data[param_position + 1]);
+		if (p_length - index >= 1) {
```
Isn't this implied by the while loop?

markofjohnson replied:
Yes but... The overall pattern in the method is that every time p_data is looked at to see if the next MIDI message can be read from it, there is this check to see if there are enough bytes for the whole message that is being checked for. I do not think that removing some of these >=1 checks  because of the while condition would improve code clarity. (I really intended to make a few necessary changes as needed to fix the bug, not rewrite the method.) I am however a beginner with how things are done at Godot.

lawnjelly commented:
It's actually decades since I did any MIDI but afaik some things to be sure here is:
* There are no infinite loops
* It  handles gracefully any unknown MIDI messages
* Ignores system exclusive messages

markofjohnson commented:
A malformed MIDI packet can cause an infinite loop hang in this code, due to my bug fix. I will make an additional commit to address this for consideration. 


> On Mar 25, 2024, at 10:44, lawnjelly ***@***.***> wrote:
> 
> 
> It's actually decades since I did any MIDI but afaik some things to be sure here is:
> 
> There are no infinite loops
> It handles gracefully any unknown MIDI messages
> Ignores system exclusive messages
> —
> Reply to this email directly, view it on GitHub <https://github.com/godotengine/godot/pull/89871#issuecomment-2018559161>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AADGCIXA7F77QIKUQX2HCTLY2BPBVAVCNFSM6AAAAABFGLMJUSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMJYGU2TSMJWGE>.
> You are receiving this because you authored the thread.
>

markofjohnson commented:
@lawnjelly and @AThousandShips are we OK with commit [d33ee9d](https://github.com/godotengine/godot/commit/d33ee9d2858d1796250287dd9983dcb9d9338eb9) re stopping the infinite loop. Adding this is necessary because I added the while loop in order to fix the bug https://github.com/godotengine/godot/issues/77035 . Its not perfect because I didn't extend the code to support every possible MIDI message, but I consider that outside the scope of fixing the original defect I ran into. 

If you approve I suppose I have to squash commits again?

AThousandShips commented:
Will take a look later today

AThousandShips commented on core/os/midi_driver.cpp:
```diff
@@ -42,77 +42,88 @@ void MIDIDriver::set_singleton() {
+				// either because its MIDI that this code does not support or a malformed MIDI packet
+				// from an upstream device. This would cause an infinite loop of the while.
+				// To prevent hanging, remaining bytes in p_data are ignored.
+				// (Which is what this method did in previous versions).
```
This message is a bit too verbose, I'd just say something like "Anything else means an invalid packet, exit.", we might want to add an error message here, unsure

lawnjelly replied:
There is `WARN_PRINT_ONCE`, but it depends whether it can get hit in non-error conditions (e.g. unrecognised event).

markofjohnson committed:
Fix missing InputEventMidi from MIDI input

MIDIDriver::receive_input_packet can be called to process an incoming midi packet that contains multiple messages. Eg playing a chord (several notes at the same time) may arrive using midi 'running status' and contain several NoteOn in one packet. This commit loops through all the messages in the midi packet creating an InputEventMidi for each one.

Includes code review feedback by A Thousand Ships <96648715+AThousandShips@users.noreply.github.com> and https://github.com/lawnjelly

Fixes #77035

Tested manually with various .mid and .sysex files to inject midi as if from midi hardware. Tested on macos with Roland keyboards. Fix isn't macos specific, but original defect 77035 found on mac, not found on windows because windows midi handling by OS 'unrolls' running status midi into individual messages before reaching Godot, which works around the defect.

markofjohnson commented:
Hey I got really lost in git and decided to start my branch over. I *think* I now have clean 1 commit with the fixes we discussed. Thanks.

lawnjelly commented:
I think you are missing some of the changes that you did earlier: e.g.:
```
				// Default case means bytes remain that are not yet turned into an InputEventMIDI,
				// either because its MIDI that this code does not support or a malformed MIDI packet
				// from an upstream device. This would cause an infinite loop of the while.
				// To prevent hanging, remaining bytes in p_data are ignored.
				// (Which is what this method did in previous versions).
				return;
```

lawnjelly commented on core/os/midi_driver.cpp:
```diff
@@ -42,77 +42,85 @@ void MIDIDriver::set_singleton() {
+				event->set_channel(0);
+				event->set_message(MIDIMessage(p_data[index]));
+				last_received_message = p_data[index];
+				index++;
```
If this is system exclusive message, will it go out of sync if the data after the first byte is not recognised?

markofjohnson replied:
I don't know, I have only enough understanding of MIDI spec and of midi_driver.cpp to fix the bug that I have reproduced.  If your question requires investigation, can that be recorded as a separate issue from this PR, as this PR is to fix defect https://github.com/godotengine/godot/issues/77035 ?

markofjohnson replied:
OK, having read what a system exclusive is https://blog.landr.com/midi-sysex/ I think you are correct. The method should return if it sees one. I don't know what the parsing behavior should be for the packet after the system exclusive. Because this PR ads the while, it should abort. Good catch! :-)

lawnjelly replied:
Well, if your PR puts this data into a read loop, so if it gets out of sync, anything could happen (it probably aborts later if not recognized, but we need to be sure of this). So if system exclusive is not being responded to it might make sense to abort here.

It's important for PR author to be confident of this (and have the answers, as reviewers might not be familiar), because we don't want to exchange one bug for another bug (or set of bugs).

lawnjelly replied:
It is literally decades since I touched MIDI streams, so I only have very rough memories.

https://www.music.mcgill.ca/~ich/classes/mumt306/StandardMIDIfileformat.html

> F0 <length> <bytes to be transmitted after F0>

> The length is stored as a variable-length quantity. It specifies the number of bytes which follow it, not including the F0 or the length itself. For instance, the transmitted message F0 43 12 00 07 F7 would be stored in a MIDI File as F0 05 43 12 00 07 F7. It is required to include the F7 at the end so that the reader of the MIDI File knows that it has read the entire message.

markofjohnson replied:
I agree. I am going to look into a unit test for MIDI packet parsing, which may mean I will be refactoring MIDIDriver::receive_input_packet into two methods. So if I disappear for a while that's what I'm doing.

markofjohnson replied:
I started a unit test for midi_driver. When `MIDIDriver::receive_input_packet` is called from a unit test, it can parse midi correctly, but when it comes to `Input::get_singleton()->parse_input_event(event);` it crashes, see below. Should unit tests not be calling deeper into the Godot Engine with `parse_input_event`? Is there some unit test setup that needs to be done to support this? I could refactor the code to separate `parse_input_event` from the midi byte parsing code so the midi parsing can be tested by itself, but this would make my PR a bigger change. @lawnjelly are you able to advise? Thanks. 

`handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.3.dev.custom_build (46bddb63b8e8f8bbba7aa4383cf4e9e868d1f5a5)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] 1   libsystem_platform.dylib            0x000000019f683584 _sigtramp + 56
[2] std::__1::recursive_mutex::lock()
[3] std::__1::recursive_mutex::lock()
[4] std::__1::unique_lock<std::__1::recursive_mutex>::unique_lock[abi:un170006](std::__1::recursive_mutex&) (in godot.macos.editor.dev.arm64) (unique_lock.h:41)
[5] std::__1::unique_lock<std::__1::recursive_mutex>::unique_lock[abi:un170006](std::__1::recursive_mutex&) (in godot.macos.editor.dev.arm64) (unique_lock.h:40)
[6] MutexLock<MutexImpl<std::__1::recursive_mutex>>::MutexLock(MutexImpl<std::__1::recursive_mutex> const&) (in godot.macos.editor.dev.arm64) (mutex.h:81)
[7] MutexLock<MutexImpl<std::__1::recursive_mutex>>::MutexLock(MutexImpl<std::__1::recursive_mutex> const&) (in godot.macos.editor.dev.arm64) (mutex.h:81)
[8] Input::parse_input_event(Ref<InputEvent> const&) (in godot.macos.editor.dev.arm64) (input.cpp:993)
[9] MIDIDriver::receive_input_packet(int, unsigned long long, unsigned char*, unsigned int) (in godot.macos.editor.dev.arm64) (midi_driver.cpp:127)`

lawnjelly replied:
That sounds like better for a separate PR imo.

markofjohnson replied:
There is a new problem though. If I flood the midi channel with many pitch change messages memory grows and the process crashes. Same test on Godot4.2 doesn't have this problem.

markofjohnson replied:
Pitch change problem solved. Parsing error from my change - pitch change is 2 data bytes. (Which original code was doing correctly.)

markofjohnson replied:
@lawnjelly This MIDI PR is ready for review. Thanks.

lawnjelly reviewed (approved):
Looks ok to me and seems to have had some testing. I don't know if we have any specific maintainers for MIDI (reduz will probably be familiar, but may be busy) but anyone else welcome to take a look. :+1: 

I don't know how many users use MIDI input so might need some testing in the wild through betas.

markofjohnson commented:
@AThousandShips this MIDI PR ready for review if you want a look. Thanks.

AThousandShips reviewed (approved):
Style and general function looks good, can't verify the validity in regards to MIDI

akien-mga commented:
CC @fbcosentino @alcomposer @voidshine @ibrahn who contributed to MIDI support in the past year or so.

fbcosentino commented:
Will try to both reproduce the issue and test the commit on both windows and linux this weekend

markofjohnson commented:
Bug has only been reproduced on Mac as far as I know. My understanding is that windows ‘unpacks’ MIDI running status into separate MIDI messages before it passes along to user processes. So the bug can’t happen on windows. Mark JOn Apr 4, 2024, at 5:06 AM, fbcosentino ***@***.***> wrote:﻿
Will try to both reproduce the issue and test the commit on both windows and linux this weekend

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you modified the open/close state.Message ID: ***@***.***>

ibrahn commented:
(Sorry I'm late)

On a read through, I see three possible issues:
- Realtime messages (0xf8 to 0xff) will interrupt running status.
- If a packet contains an "unsupported" message type, the remainder of the packet will be discarded.
- Messages from one device can interrupt running status on another. The status bit cache `last_received_message` should be per connection, rather than global.

I think these problems will already have existed for MacOS builds, so this PR still improves things.

It looks like #77035 affects only Core MIDI (MacOS).
We currently only have two other MIDI drivers:
- on Windows, the platform API handles running status for us and drops SysEx.
- on Linux, our ALSA MIDI driver does this parsing, before passing single complete messages to `receive_input_packet`.

So the argument could be made to fix this at the [Core MIDI driver](https://github.com/godotengine/godot/tree/master/drivers/coremidi).
However, looking forward it probably does makes sense to have a parser at the platform independent level, making it less work to add MIDI input drivers for iOS, Android etc.
Perhaps we should move the [existing implementation](https://github.com/godotengine/godot/blob/a7b860250f305f6cbaf61c30f232ff3bbdfdda0b/drivers/alsamidi/midi_driver_alsamidi.cpp#L84) up from the ALSA class to the platform independent. I can look at that this afternoon/tomorrow.

markofjohnson commented:
> I think these problems will already have existed for MacOS builds, so this PR still improves things.

Yes, and cited are advanced cases of MIDI usage that would require more Godot development to support anyway. Without a fix, MIDI on Mac only works for single note at a time. BTW I tested MIDI in with 2 devices and indeed it's possible to corrupt running status data. Result was bad data through to the gdscript code but no crash. 

> So the argument could be made to fix this at the [Core MIDI driver]

Would require duplicate parsing code in midi_driver_coremidi.cpp, and it would have to allocate memory and copy data in order to turn running status into individual messages? Not ideal but it would isolate this fix from Windows.

> Looking forward it probably does makes sense to have a parser at the platform independent level, making it less work to add MIDI input drivers for iOS, Android etc. 

Good idea, especially working towards MIDI in for mobile, but out of scope for this bugfix PR. 

thanks.

ibrahn commented:
I've done the MIDI parser move, over at #90485. Could you run your chord test against it? I'm afraid I don't have access to a Mac at the moment

markofjohnson commented:
> I've done the MIDI parser move, over at #90485. Could you run your chord test against it? I'm afraid I don't have access to a Mac at the moment

I tested chord playing on Mac. This fixes the bug :-) . I logged data going in and confirmed that packets with more than one MIDI message work correctly. I noticed that I was _not_ seeing MIDI running status, but packets with multiple messages. (So the previous diagnosis of the problem may not have been due to running status, but due to more than 1 MIDI message in 1 packet.)

ibrahn commented:
I'm not sure what the process is at the moment, but I'd guess leave it open for now, at least until we know if #90485 passes review. As a fallback to keep MacOS MIDI on the cards for 4.3

akien-mga commented:
I believe this is superseded by #90485, but I'm not too familiar with the details. Are there still changes from this PR which would be worth merge after rebasing / redoing the changes, or should this be closed now?

markofjohnson commented:
withdrawing pr because https://github.com/godotengine/godot/pull/90485

markofjohnson commented:
Superseded by https://github.com/godotengine/godot/pull/90485

akien-mga commented:
Thanks for the contribution anyway, which led to finding the better cross-platform solution.
